## **Business Understanding**
Performing Analysis on this information

**About Data:**
Concrete is the most important material in civil engineering. The concrete compressive strength is a highly nonlinear function of age and ingredients.


**Reference**
Yeh, I. (1998). Concrete Compressive Strength [Dataset]. UCI Machine Learning Repository. https://doi.org/10.24432/C5PK67.

#### **Loading Libraries**

In [1]:
import pandas as pd 

#### **Loading data**

In [2]:
df = pd.read_excel("data/concrete+compressive+strength/Concrete_Data.xls") # reads the data and creates a dataframe(2D or 1D- series represntation of your data) from it

## **Data Understanding**

#### **Visual perspective of the data**

- top 5
- top 10
- last 5
- last 10
- shape
- info
- duplicates

In [3]:
df.head() # frist 5 records

,Cement (component 1)(kg in a m^3 mixture),Blast Furnace Slag (component 2)(kg in a m^3 mixture),Fly Ash (component 3)(kg in a m^3 mixture),Water (component 4)(kg in a m^3 mixture),Superplasticizer (component 5)(kg in a m^3 mixture),Coarse Aggregate (component 6)(kg in a m^3 mixture),Fine Aggregate (component 7)(kg in a m^3 mixture),Age (day),"Concrete compressive strength(MPa, megapascals)"
0,540.0,0.0,0.0,162.0,2.5,1040.0,676.0,28,79.986111
1,540.0,0.0,0.0,162.0,2.5,1055.0,676.0,28,61.887366
2,332.5,142.5,0.0,228.0,0.0,932.0,594.0,270,40.269535
3,332.5,142.5,0.0,228.0,0.0,932.0,594.0,365,41.052780
4,198.6,132.4,0.0,192.0,0.0,978.4,825.5,360,44.296075


In [4]:
df.head(10) # first ten records

,Cement (component 1)(kg in a m^3 mixture),Blast Furnace Slag (component 2)(kg in a m^3 mixture),Fly Ash (component 3)(kg in a m^3 mixture),Water (component 4)(kg in a m^3 mixture),Superplasticizer (component 5)(kg in a m^3 mixture),Coarse Aggregate (component 6)(kg in a m^3 mixture),Fine Aggregate (component 7)(kg in a m^3 mixture),Age (day),"Concrete compressive strength(MPa, megapascals)"
0,540.0,0.0,0.0,162.0,2.5,1040.0,676.0,28,79.986111
1,540.0,0.0,0.0,162.0,2.5,1055.0,676.0,28,61.887366
2,332.5,142.5,0.0,228.0,0.0,932.0,594.0,270,40.269535
3,332.5,142.5,0.0,228.0,0.0,932.0,594.0,365,41.052780
4,198.6,132.4,0.0,192.0,0.0,978.4,825.5,360,44.296075
5,266.0,114.0,0.0,228.0,0.0,932.0,670.0,90,47.029847
6,380.0,95.0,0.0,228.0,0.0,932.0,594.0,365,43.698299
7,380.0,95.0,0.0,228.0,0.0,932.0,594.0,28,36.447770
8,266.0,114.0,0.0,228.0,0.0,932.0,670.0,28,45.854291
9,475.0,0.0,0.0,228.0,0.0,932.0,594.0,28,39.289790


In [5]:
df.tail() # last five records

,Cement (component 1)(kg in a m^3 mixture),Blast Furnace Slag (component 2)(kg in a m^3 mixture),Fly Ash (component 3)(kg in a m^3 mixture),Water (component 4)(kg in a m^3 mixture),Superplasticizer (component 5)(kg in a m^3 mixture),Coarse Aggregate (component 6)(kg in a m^3 mixture),Fine Aggregate (component 7)(kg in a m^3 mixture),Age (day),"Concrete compressive strength(MPa, megapascals)"
1025,276.4,116.0,90.3,179.6,8.9,870.1,768.3,28,44.284354
1026,322.2,0.0,115.6,196.0,10.4,817.9,813.4,28,31.178794
1027,148.5,139.4,108.6,192.7,6.1,892.4,780.0,28,23.696601
1028,159.1,186.7,0.0,175.6,11.3,989.6,788.9,28,32.768036
1029,260.9,100.5,78.3,200.6,8.6,864.5,761.5,28,32.401235


In [6]:
df.tail(10) # last ten records

,Cement (component 1)(kg in a m^3 mixture),Blast Furnace Slag (component 2)(kg in a m^3 mixture),Fly Ash (component 3)(kg in a m^3 mixture),Water (component 4)(kg in a m^3 mixture),Superplasticizer (component 5)(kg in a m^3 mixture),Coarse Aggregate (component 6)(kg in a m^3 mixture),Fine Aggregate (component 7)(kg in a m^3 mixture),Age (day),"Concrete compressive strength(MPa, megapascals)"
1020,288.4,121.0,0.0,177.4,7.0,907.9,829.5,28,42.140084
1021,298.2,0.0,107.0,209.7,11.1,879.6,744.2,28,31.875165
1022,264.5,111.0,86.5,195.5,5.9,832.6,790.4,28,41.542308
1023,159.8,250.0,0.0,168.4,12.2,1049.3,688.2,28,39.455954
1024,166.0,259.7,0.0,183.2,12.7,858.8,826.8,28,37.917043
1025,276.4,116.0,90.3,179.6,8.9,870.1,768.3,28,44.284354
1026,322.2,0.0,115.6,196.0,10.4,817.9,813.4,28,31.178794
1027,148.5,139.4,108.6,192.7,6.1,892.4,780.0,28,23.696601
1028,159.1,186.7,0.0,175.6,11.3,989.6,788.9,28,32.768036
1029,260.9,100.5,78.3,200.6,8.6,864.5,761.5,28,32.401235


In [8]:
rows, columns = df.shape
output = f"{rows} records and {columns} features"
print(output)

1030 records and 9 features


<hr>

#### This is fo illustartion purposes to explain or show null data concept i.e null values/ missing values

In [14]:
fish_df= pd.read_csv("data/Untitled form (Responses) - Form Responses 1.csv")

In [15]:
fish_df.shape

(4, 3)

In [16]:
fish_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 3 columns):
 #   Column                                              Non-Null Count  Dtype
---  ------                                              --------------  -----
 0   Timestamp                                           4 non-null      str  
 1   Which fish type do you prefer?                      4 non-null      str  
 2   Which part of Nairobi you prefere to get it from:   1 non-null      str  
dtypes: str(3)
memory usage: 228.0 bytes


<hr>

**Null values:** 
- missing values in a dataset 

Remedy: 
    - if the **missing values** are more than 80% - drop the column (delete or exclude the column from your analysis)
    - if the **missing values** are less than 20% - we drop the rows containing the missing values
    - if the **missing values** are above 40% or 50% -  we conduct what we call **imputation** (we fill in the missing values using eg an MCT)

    Kindly note:
    - [About missing values](https://towardsdatascience.com/all-about-missing-data-handling-b94b8b5d2184/)
    - [About missing values](https://www-users.york.ac.uk/~mb55/intro/typemiss4.htm)

In [10]:
df.info() # tells us  count of non-null values and data types

<class 'pandas.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column                                                 Non-Null Count  Dtype  
---  ------                                                 --------------  -----  
 0   Cement (component 1)(kg in a m^3 mixture)              1030 non-null   float64
 1   Blast Furnace Slag (component 2)(kg in a m^3 mixture)  1030 non-null   float64
 2   Fly Ash (component 3)(kg in a m^3 mixture)             1030 non-null   float64
 3   Water  (component 4)(kg in a m^3 mixture)              1030 non-null   float64
 4   Superplasticizer (component 5)(kg in a m^3 mixture)    1030 non-null   float64
 5   Coarse Aggregate  (component 6)(kg in a m^3 mixture)   1030 non-null   float64
 6   Fine Aggregate (component 7)(kg in a m^3 mixture)      1030 non-null   float64
 7   Age (day)                                              1030 non-null   int64  
 8   Concrete compressive strength(MPa, megapascals)        1030 

In [18]:
df.dtypes # returms the data type of each column

Cement (component 1)(kg in a m^3 mixture)                float64
Blast Furnace Slag (component 2)(kg in a m^3 mixture)    float64
Fly Ash (component 3)(kg in a m^3 mixture)               float64
Water  (component 4)(kg in a m^3 mixture)                float64
Superplasticizer (component 5)(kg in a m^3 mixture)      float64
Coarse Aggregate  (component 6)(kg in a m^3 mixture)     float64
Fine Aggregate (component 7)(kg in a m^3 mixture)        float64
Age (day)                                                  int64
Concrete compressive strength(MPa, megapascals)          float64
dtype: object

In [19]:
df.isnull().sum() # check missing values

Cement (component 1)(kg in a m^3 mixture)                0
Blast Furnace Slag (component 2)(kg in a m^3 mixture)    0
Fly Ash (component 3)(kg in a m^3 mixture)               0
Water  (component 4)(kg in a m^3 mixture)                0
Superplasticizer (component 5)(kg in a m^3 mixture)      0
Coarse Aggregate  (component 6)(kg in a m^3 mixture)     0
Fine Aggregate (component 7)(kg in a m^3 mixture)        0
Age (day)                                                0
Concrete compressive strength(MPa, megapascals)          0
dtype: int64

In [21]:
df.duplicated().sum() # the  umber of duplicate data

np.int64(25)

In [36]:
features = df.columns # returns the column names as a list

In [37]:
features

Index(['Cement (component 1)(kg in a m^3 mixture)',
       'Blast Furnace Slag (component 2)(kg in a m^3 mixture)',
       'Fly Ash (component 3)(kg in a m^3 mixture)',
       'Water  (component 4)(kg in a m^3 mixture)',
       'Superplasticizer (component 5)(kg in a m^3 mixture)',
       'Coarse Aggregate  (component 6)(kg in a m^3 mixture)',
       'Fine Aggregate (component 7)(kg in a m^3 mixture)', 'Age (day)',
       'Concrete compressive strength(MPa, megapascals) '],
      dtype='str')

In [39]:
features[3] # using index to access elemnts in the list our list is a list of column names

'Water  (component 4)(kg in a m^3 mixture)'

In [29]:
df.duplicated(subset=[features[3]]).any()

np.True_

In [35]:
for feature in features:
    print(f"=> {feature}  <{df.duplicated(subset=[features[3]]).any()}>")

=> Cement (component 1)(kg in a m^3 mixture)  <True>
=> Blast Furnace Slag (component 2)(kg in a m^3 mixture)  <True>
=> Fly Ash (component 3)(kg in a m^3 mixture)  <True>
=> Water  (component 4)(kg in a m^3 mixture)  <True>
=> Superplasticizer (component 5)(kg in a m^3 mixture)  <True>
=> Coarse Aggregate  (component 6)(kg in a m^3 mixture)  <True>
=> Fine Aggregate (component 7)(kg in a m^3 mixture)  <True>
=> Age (day)  <True>
=> Concrete compressive strength(MPa, megapascals)   <True>


In [34]:
df.shape

(1030, 9)

#### **Measures of Central Tendency**
- Attaining one numerical value that is represtative of the entire data set( feature).  
     - MEAN(arithmetic mean)
     - MEDIAN(after arraing your data either in ascending or descending order you get the center item) 
     - MODE(most frequently repeated value)
     

In [46]:
mean = df["Cement (component 1)(kg in a m^3 mixture)"].mean()
mean

np.float64(281.16563106796116)

In [47]:
sd = df.std()
sd

Cement (component 1)(kg in a m^3 mixture)                104.507142
Blast Furnace Slag (component 2)(kg in a m^3 mixture)     86.279104
Fly Ash (component 3)(kg in a m^3 mixture)                63.996469
Water  (component 4)(kg in a m^3 mixture)                 21.355567
Superplasticizer (component 5)(kg in a m^3 mixture)        5.973492
Coarse Aggregate  (component 6)(kg in a m^3 mixture)      77.753818
Fine Aggregate (component 7)(kg in a m^3 mixture)         80.175427
Age (day)                                                 63.169912
Concrete compressive strength(MPa, megapascals)           16.705679
dtype: float64

In [49]:
df["Cement (component 1)(kg in a m^3 mixture)"].describe()

count    1030.000000
mean      281.165631
std       104.507142
min       102.000000
25%       192.375000
50%       272.900000
75%       350.000000
max       540.000000
Name: Cement (component 1)(kg in a m^3 mixture), dtype: float64

In [48]:
df.describe()

,Cement (component 1)(kg in a m^3 mixture),Blast Furnace Slag (component 2)(kg in a m^3 mixture),Fly Ash (component 3)(kg in a m^3 mixture),Water (component 4)(kg in a m^3 mixture),Superplasticizer (component 5)(kg in a m^3 mixture),Coarse Aggregate (component 6)(kg in a m^3 mixture),Fine Aggregate (component 7)(kg in a m^3 mixture),Age (day),"Concrete compressive strength(MPa, megapascals)"
count,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000
mean,281.165631,73.895485,54.187136,181.566359,6.203112,972.918592,773.578883,45.662136,35.817836
std,104.507142,86.279104,63.996469,21.355567,5.973492,77.753818,80.175427,63.169912,16.705679
min,102.000000,0.000000,0.000000,121.750000,0.000000,801.000000,594.000000,1.000000,2.331808
25%,192.375000,0.000000,0.000000,164.900000,0.000000,932.000000,730.950000,7.000000,23.707115
50%,272.900000,22.000000,0.000000,185.000000,6.350000,968.000000,779.510000,28.000000,34.442774
75%,350.000000,142.950000,118.270000,192.000000,10.160000,1029.400000,824.000000,56.000000,46.136287
max,540.000000,359.400000,200.100000,247.000000,32.200000,1145.000000,992.600000,365.000000,82.599225


#### **Univariate Analysis**
- analysis of single features

In [50]:
df["Cement (component 1)(kg in a m^3 mixture)"].describe()

count    1030.000000
mean      281.165631
std       104.507142
min       102.000000
25%       192.375000
50%       272.900000
75%       350.000000
max       540.000000
Name: Cement (component 1)(kg in a m^3 mixture), dtype: float64

In [51]:
IQR = df["Cement (component 1)(kg in a m^3 mixture)"].quantile(0.75) - df["Cement (component 1)(kg in a m^3 mixture)"].quantile(0.25) 
IQR

np.float64(157.625)

In [52]:
df_range = df["Cement (component 1)(kg in a m^3 mixture)"].max() - df["Cement (component 1)(kg in a m^3 mixture)"].min()
df_range

np.float64(438.0)

In [53]:
df_variance = df["Cement (component 1)(kg in a m^3 mixture)"].var()
df_variance

np.float64(10921.742654363268)

In [54]:
df_sd = df["Cement (component 1)(kg in a m^3 mixture)"].std()
df_sd

np.float64(104.5071416428718)

In [ ]:
df_median = df["Cement (component 1)(kg in a m^3 mixture)"].median()
df_median

np.float64(272.9)

In [55]:
df_mode = df["Cement (component 1)(kg in a m^3 mixture)"].mode()
df_mode

0    362.6
1    425.0
Name: Cement (component 1)(kg in a m^3 mixture), dtype: float64

In [58]:
# (362 + 425)/3 # TODO: confirm mathematical formular + rational

262.3333333333333

**Insights**
- We have 1030 records
- most m^3 mixtures were mixed with an average of 281kgs of component 1
- the spread of the data from the measure of central tendency 
(Mean, Median, Mode) is about 104kgs
- 102kg was the lowest mixtire quantity of component 1
- 25% of the mixture quantities were below 192.4kgs
- 50% of the mixtire levels were at 272kgs (can be used as the Median)
- 75% of the mixture levels were below 350kgs 
- highest mixtire recorded was at 540kgs 
- majority of the mixtures fell around 157.625kgs

#### **Bivariate Analysis**
- analysis of two features

#### **Multivariate Analysis**
- analysis of more than two features

## **Data Preparation**

In [43]:
clean_df = df.drop_duplicates() # drops all duplicated rows (kinldy note some instances of duplicates could be the actual truth on ground)
clean_df.shape

(1005, 9)

## **Modeling**

## **Evaluation**

## **Deployment**